sygnal/helper/context_factory.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2014-2016 OpenMarket Ltd
+# Copyright 2019 New Vector Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Adapted from Synapse:
+#  https://github.com/matrix-org/synapse/blob/1016f303e58b1305ed5b3572fde002e1273e0fc0/synapse/crypto/context_factory.py#L77
+
+
+import logging
+
+import idna
+from OpenSSL import SSL
+from service_identity import VerificationError
+from service_identity.pyopenssl import verify_hostname, verify_ip_address
+from twisted.internet.abstract import isIPAddress, isIPv6Address
+from twisted.internet.interfaces import IOpenSSLClientConnectionCreator
+from twisted.internet.ssl import CertificateOptions, TLSVersion, platformTrust
+from twisted.python.failure import Failure
+from twisted.web.iweb import IPolicyForHTTPS
+from zope.interface import implementer
+
+logger = logging.getLogger(__name__)
+
+
+@implementer(IPolicyForHTTPS)
+class ClientTLSOptionsFactory(object):
+    """Factory for Twisted SSLClientConnectionCreators that are used to make connections
+    to remote servers for federation.
+    Uses one of two OpenSSL context objects for all connections, depending on whether
+    we should do SSL certificate verification.
+    get_options decides whether we should do SSL certificate verification and
+    constructs an SSLClientConnectionCreator factory accordingly.
+    """
+
+    def __init__(self):
+        # Use CA root certs provided by OpenSSL
+        trust_root = platformTrust()
+
+        # "insecurelyLowerMinimumTo" is the argument that will go lower than
+        # Twisted's default, which is why it is marked as "insecure" (since
+        # Twisted's defaults are reasonably secure). But, since Twisted is
+        # moving to TLS 1.2 by default, we want to respect the config option if
+        # it is set to 1.0 (which the alternate option, raiseMinimumTo, will not
+        # let us do).
+        minTLS = TLSVersion.TLSv1_2
+
+        self._verify_ssl = CertificateOptions(
+            trustRoot=trust_root, insecurelyLowerMinimumTo=minTLS
+        )
+        self._verify_ssl_context = self._verify_ssl.getContext()
+        self._verify_ssl_context.set_info_callback(self._context_info_cb)
+
+        self._no_verify_ssl = CertificateOptions(insecurelyLowerMinimumTo=minTLS)
+        self._no_verify_ssl_context = self._no_verify_ssl.getContext()
+        self._no_verify_ssl_context.set_info_callback(self._context_info_cb)
```

Comment:
Let's just rip out all this `no_verify` stuff and just have a single `_verify_ssl_context`